File: samples/scoping-samples/simple-stack-example-scoping-java/src/main/java/com/zhuinden/simplestackexamplescoping/features/words/WordListFragment.java
package com.zhuinden.simplestackexamplescoping.features.words;

import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import com.zhuinden.eventemitter.EventSource;
import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.navigator.Navigator;
import com.zhuinden.simplestackexamplescoping.R;
import com.zhuinden.simplestackextensions.fragments.KeyedFragment;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;

public class WordListFragment
        extends KeyedFragment {
    public WordListFragment() {
        super(R.layout.word_list_view);
    }

    public interface ActionHandler {
        public void onAddNewWordClicked();
    }

    public interface DataProvider {
        public Observable<List<String>> getWordList();
    }

    private ActionHandler actionHandler;

    private DataProvider dataProvider;
    private EventSource<WordController.Events> controllerEvents;

    private WordListAdapter adapter = new WordListAdapter();

    private EventSource.NotificationToken notificationToken = null;

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final Backstack backstack = Navigator.getBackstack(requireContext());
        dataProvider = backstack.lookupService(DataProvider.class.getName());
        actionHandler = backstack.lookupService(ActionHandler.class.getName());
        controllerEvents = backstack.lookupService("WordEventEmitter");

        RecyclerView recyclerView = view.findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false));
        recyclerView.setAdapter(adapter);

        view.findViewById(R.id.buttonGoToAddNewWord).setOnClickListener(v -> actionHandler.onAddNewWordClicked());


    }

    @Override
    public void onStart() {
        super.onStart();

        notificationToken = controllerEvents.startListening(event -> {
            if(event instanceof WordController.Events.NewWordAdded) {
                Toast.makeText(requireContext(),
                               "Added " + ((WordController.Events.NewWordAdded) event).getWord(),
                               Toast.LENGTH_SHORT).show();
            }
        });

        compositeDisposable.add(dataProvider.getWordList().subscribe(strings -> adapter.updateWords(strings)));
    }

    @Override
    public void onStop() {
        compositeDisposable.clear();

        if(notificationToken != null) {
            notificationToken.stopListening();
            notificationToken = null;
        }

        super.onStop();
    }
}

File: samples/scoping-samples/simple-stack-example-scoping/src/main/java/com/zhuinden/simplestackexamplescoping/features/words/WordListFragment.kt
package com.zhuinden.simplestackexamplescoping.features.words

import android.os.Bundle
import android.view.View
import androidx.lifecycle.LiveData
import androidx.lifecycle.observe
import androidx.recyclerview.widget.LinearLayoutManager
import com.zhuinden.liveevent.observe
import com.zhuinden.simplestackexamplescoping.R
import com.zhuinden.simplestackexamplescoping.databinding.WordListViewBinding
import com.zhuinden.simplestackexamplescoping.utils.onClick
import com.zhuinden.simplestackexamplescoping.utils.safe
import com.zhuinden.simplestackexamplescoping.utils.showToast
import com.zhuinden.simplestackexamplescoping.utils.viewBinding
import com.zhuinden.simplestackextensions.fragments.KeyedFragment
import com.zhuinden.simplestackextensions.fragmentsktx.lookup

/**
 * Created by Zhuinden on 2018.09.17.
 */

class WordListFragment : KeyedFragment(R.layout.word_list_view) {
    private val binding by viewBinding(WordListViewBinding::bind)

    interface ActionHandler {
        fun onAddNewWordClicked()
    }

    interface DataProvider {
        val wordList: LiveData<List<String>>
    }

    private val actionHandler by lazy { lookup<ActionHandler>() }
    private val dataProvider by lazy { lookup<DataProvider>() }
    private val wordList by lazy { dataProvider.wordList }

    private val controllerEvents by lazy { lookup<WordEventEmitter>() }

    val adapter = WordListAdapter()

    @Suppress("NAME_SHADOWING")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        binding.recyclerView.adapter = adapter

        binding.buttonGoToAddNewWord.onClick { view ->
            actionHandler.onAddNewWordClicked()
        }

        wordList.observe(viewLifecycleOwner) { words ->
            adapter.updateWords(words)
        }

        controllerEvents.observe(viewLifecycleOwner) { event ->
            when (event) {
                is WordController.Events.NewWordAdded -> showToast("Added ${event.word}")
            }.safe()
        }
    }
}